atlasdb-tests-shared/src/test/java/com/palantir/atlasdb/transaction/impl/SnapshotTransactionTest.java

```diff
@@ -1684,6 +1692,185 @@ public void orphanedSentinelsDoNotGetDeletedAfterDetectionInConservativelySweptT
+
+        Iterator<Map.Entry<Cell, byte[]>> sortedColumns = transaction.getSortedColumns(
+                TABLE_SWEPT_THOROUGH,
+                rows,
+                BatchColumnRangeSelection.create(PtBytes.EMPTY_BYTE_ARRAY, PtBytes.EMPTY_BYTE_ARRAY, 1));
+        assertThat(sortedColumns.next().getKey()).isEqualTo(cells.get(0));
+
+        // lock lost after getting first batch
+        timelockService.unlock(ImmutableSet.of(res.getLock()));
+        assertThatThrownBy(sortedColumns::next).isInstanceOf(TransactionLockTimeoutException.class);
+    }
+
+    @Test
+    public void getSortedColumnsObeysColumnRangeSelection() {
+        List<byte[]> rows = ImmutableList.of(ROW_FOO, ROW_BAR);
+        List<Cell> col1_cells = ImmutableList.of(Cell.create(ROW_FOO, COL_A), Cell.create(ROW_BAR, COL_A));
+
+        putCellsInTable(col1_cells, TABLE);
+        List<Cell> entries =
+                getSortedEntries(TABLE, rows, BatchColumnRangeSelection.create(COL_A, "az".getBytes(), 1000));
+        Assertions.assertThat(entries).containsExactlyElementsOf(col1_cells);
+
+        List<Cell> outOfRangeEntries =
+                getSortedEntries(TABLE, rows, BatchColumnRangeSelection.create("y".getBytes(), "z".getBytes(), 1000));
+        Assertions.assertThat(outOfRangeEntries).isEmpty();
+    }
+
+    @Test
+    public void getSortedColumnsIteratesOverMultipleBatchesIfRequired() {
+        ImmutableList<byte[]> rows = ImmutableList.of(ROW_FOO, ROW_BAR);
+        ImmutableList<byte[]> columns = ImmutableList.of(COL_A, COL_B);
+
+        List<Cell> cells = rows.stream()
+                .map(row -> columns.stream().map(col -> Cell.create(row, col)).collect(Collectors.toList()))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+
+        putCellsInTable(cells, TABLE);
+        List<Cell> entries = getSortedEntries(
+                TABLE, rows, BatchColumnRangeSelection.create(PtBytes.EMPTY_BYTE_ARRAY, PtBytes.EMPTY_BYTE_ARRAY, 1));
+
+        List<Cell> entriesSortedByColumn = columns.stream()
+                .map(col -> rows.stream().map(row -> Cell.create(row, col)).collect(Collectors.toList()))
+                .flatMap(Collection::stream)
+                .collect(Collectors.toList());
+        Assertions.assertThat(entries).containsExactlyElementsOf(entriesSortedByColumn);
+    }
+
+    @Test
+    public void getSortedColumnsValidatesLocksOncePerBatch() {
+        List<byte[]> rows = LongStream.range(0, 1000).mapToObj(PtBytes::toBytes).collect(Collectors.toList());
+
+        List<Cell> cells = rows.stream().map(row -> Cell.create(row, COL_A)).collect(Collectors.toList());
+        putCellsInTable(cells, TABLE_SWEPT_THOROUGH);
+
+        verifyPrefetchValidations(rows, cells, 100, 10, 1000);
+        verifyPrefetchValidations(rows, cells, 100, 3, 299);
+        verifyPrefetchValidations(rows, cells, 100, 4, 300);
+        verifyPrefetchValidations(rows, cells, 100, 4, 301);
+    }
+
```

Comment:
I think we are still missing one set of test cases, though they are likely to be a little annoying so I would prefer for them to be in a separate PR. Specifically, the logic of how we deal with wide rows is not tested yet. So something like this is missing:

```
    @Test
    public void getSortedColumnsAvoidsVeryLargeReadsInWideRows() {
        List<byte[]> rows = Stream.of("tom", "jerry", "alice", "bob", "cari", "david")
                .map(PtBytes::toBytes)
                .collect(Collectors.toList());
        List<byte[]> columns = LongStream.range(0, 10_000).mapToObj(PtBytes::toBytes).collect(Collectors.toList());

        // Only provide entries for the tom row
        List<Cell> cells = columns.stream().map(column -> Cell.create(
                rows.get(0), column)).collect(Collectors.toList());
        putCellsInTable(cells, TABLE_SWEPT_THOROUGH);

        verifyPrefetchValidations(
                rows, cells, 10_000, 1, 10_000);
        // verify that any gRCR on the KVS has a batch hint of *strictly less* than 10k
    }
```

Also: is the above result what you expect, and is it correct?